gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -213,16 +262,62 @@ public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedE
     return isTerminated();
   }
 
+  /** Scheduling loop. */
+  private ScheduledFuture<?> scheduleNextRefresh() {
+    long delayPeriod = REFRESH_PERIOD.toMillis();
+    long jitter = (long) ((Math.random() - 0.5) * JITTER_PERCENTAGE * delayPeriod);
+    long delay = jitter + delayPeriod;
+    return channelRefreshExecutorService.schedule(
+        () -> {
+          try {
+            refresh();
+          } finally {
+            scheduleNextRefresh();
+          }
+        },
+        delay,
+        TimeUnit.MILLISECONDS);
+  }
+
   /**
-   * Performs a simple round robin on the list of {@link ManagedChannel}s in the {@code channels}
-   * list.
+   * Replace all of the channels in the channel pool with fresh ones. This is meant to mitigate the
+   * hourly GFE disconnects by giving clients the ability to prime the channel on reconnect.
    *
-   * @return A {@link ManagedChannel} that can be used for a single RPC call.
+   * <p>This is done on a best effort basis. If the replacement channel fails to construct, the old
+   * channel will continue to be used.
```

Comment:
Add a comment like "the method is not thread-safe and must be called serially"?

Reply:
I just updated the code to make sure its thread safe

Reply:
I am fine with leaving the code not thread-safe but just say that no one should call this method concurrently for the sake of simplicity. I see that `refresh()` is intended to be called only by the refresh scheduler periodically, and indeed it's never called concurrently as of now.

Why I am saying this is that, calling `refresh()` concurrently is still problematic even after the change, and I start to doubt if it's really worth making the code more complex.

For example, let's assume that `localEntries.size() == 1` for simplicity, and call the sole element in the list "Entry O" (original Entry).

Thread A & B: both enter `refresh()` at the same time.
Thread A: creates a new `Entry`, say "Entry N" (new Entry). By the time it reaches line 294, `newEntries == [Entry N]` and `removeEntries = [Entry O]`.
Thread B: `new Entry(...)` throws `IOException`. In this case, nothing happens or changes: `newEntries == [Entry O]` and `removeEntries` is empty by the time it reaches line 294.

Thread A: `entries.getAndSet()` makes `entries == [Entry N]`. Nothing interesting happens in the following loop, and `requestShutdown()` is called on "Entry O."
Thread B: calls `entries.getAndSet()`, whose effect is to revert `entries` back to `[Entry O]`. However, "Entry O" is already shut down by Thread A.

Reply:
I'd like to try to keep this method thread safe. I re-arranged the code to account for concurrency without introducing a lot complexity. PTAL

Reply:
I wouldn't mind a simple approach to synchronize even the whole method. I think everything will be super fast anyway, not to mention that `refresh()` is never called concurrently in practice.

Reply:
Another really good catch. I added a monitor to address the problem. Thanks again for your indepth review!